crates/ra_ide/src/completion/complete_unqualified_path.rs

```diff
@@ -4,17 +4,42 @@ use hir::ScopeDef;
 use test_utils::tested_by;
 
 use crate::completion::{CompletionContext, Completions};
-use ra_syntax::AstNode;
+use hir::Adt;
+use ra_syntax::SyntaxNode;
+use ra_syntax::{ast, AstNode, SyntaxKind};
 
 pub(super) fn complete_unqualified_path(acc: &mut Completions, ctx: &CompletionContext) {
-    if !ctx.is_trivial_path {
+    if ctx.record_lit_syntax.is_some() || ctx.record_pat_syntax.is_some() {
         return;
     }
 
-    if ctx.is_pat_binding_or_const
-        || ctx.record_lit_syntax.is_some()
-        || ctx.record_pat_syntax.is_some()
-    {
+    for ancestor in ctx.token.ancestors() {
+        if let Some(()) = match ancestor.kind() {
```

Comment:
I don't understand why this should be necessary -- why not just look for a `Pat` in the ancestors?

Also, I'd suggest moving this computation of the 'expected type' to `CompletionContext`.

Reply:
Not sure why I didn't think to try that, it seems obvious in retrospect! That also simplifies the code greatly.